openedx_external_enrollments/external_enrollments/pathstream_external_enrollment.py

```diff
@@ -0,0 +1,201 @@
+                self.S3_FILE
+            )
+
+        except ClientError as error:
+            LOG.error('Failed to download the file to S3. Reason: %s', str(error))
+            return False
+
+        LOG.info('File [%s] was downloaded from S3 for [%s]', self.S3_FILE, self.__str__())
+        return True
+
+    def _update_file(self, enrollments_data):
+        """Append enrollment data to the downloaded S3 file
+
+        :param enrollments_data: List of the enrollments data
+        """
+        if os.path.exists(self.S3_FILE):
+
+            with open(self.S3_FILE, 'a') as f:
+                f.writelines(enrollments_data)
+        else:
+            LOG.error('File does not exist or _download_file has not been called')
+
+    def _upload_file(self):
+        """Upload the file to an S3 bucket
+
+        :param file_name: File to upload
+        :param bucket: Bucket to upload to
+
+        :return: True if file was uploaded, else False
+        """
+        if self.client is None:
+            raise S3NotInitialized
+
+        try:
+            self.client.upload_file(
+                self.S3_FILE,
+                self.S3_BUCKET,
+                self.S3_FILE
+            )
+
+        except ClientError as error:
+            LOG.error('Failed to upload the file to S3. Reason: %s', str(error))
+            return False
+
+        LOG.info('File [%s] was uploaded to S3 for [%s]', self.S3_FILE, self.__str__())
+        return True
+
+    def _delete_downloaded_file(self):
+        """Delete the downloaded file.
+        """
+        if os.path.exists(self.S3_FILE):
+            os.remove(self.S3_FILE)
+        else:
+            LOG.error('File could not be deleted, File does not exist')
+
+    def execute_upload(self):
+        """
+        Context: This method will download, update and upload an S3 File with the
+        corresponding data from new enrollments which have not been uploaded to the S3
+        file.
+
+        :return: True if completed, else False.
+
+        Complete means that this method either download-update-upload the S3 file or do not
+        proceed to download-update-upload the file because there are no new enrollments.
+        """
+
+        enrollments_qs = ExternalEnrollment.objects.filter(  # pylint: disable=no-member
+            controller_name=str(self),
+            meta__icontains='"is_uploaded":false',
+        )
+
+        if enrollments_qs.count() > 0:
+
+            self._init_s3()
+
+            is_file_downloaded = self._download_file()
+
+            if is_file_downloaded:
+
+                enrollments_data = [enrollment.meta['enrollment_data_formated'] for enrollment in enrollments_qs]
+
+                self._update_file(enrollments_data)
+
+                is_uploaded = self._upload_file()
+
+                if is_uploaded:
+
+                    for e in enrollments_qs:
+                        e.meta.update({'is_uploaded': True})
+                        e.save()
```

Comment:
@Jacatove, probably you can use [`bulk_update`](https://docs.djangoproject.com/en/2.2/ref/models/querysets/#bulk-update) for this, could you check that out please?

Reply:
@ivanvgh  Yes It worked and the docs say is more efficient